More guidance on schema dialects. (#3909)
OAI/OpenAPI-Specification: versions/3.1.1.md

In addition to the JSON Schema properties comprising the OAS dialect, the Schema Object supports keywords from any other vocabularies, or entirely arbitrary properties.

> The OpenAPI Specification's base vocabulary is safe for standard validators to ignore, as indicated by its inclusion in the OAS dialect's [`$vocabulary`](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-8.1.2) keyword with a value of `false`.

Reviewer: ```suggestion
Validators MAY choose to ignore keywords defined by the OpenAPI Specification's base vocabulary, due to its inclusion in the OAS dialect with a [`$vocabulary`](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-8.1.2) value of `false`.
```

It is important for tooling to be able to determine which dialect or meta-schema any given resource wishes to be processed with: JSON Schema Core, JSON Schema Validation, OpenAPI Schema dialect, or some custom meta-schema.

The `$schema` keyword MAY be present in any Schema Object that is a [schema resource root](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-4.3.5), and if present MUST be used to determine which dialect should be used when processing the schema. This allows use of Schema Objects which comply with other drafts of JSON Schema than the default Draft 2020-12 support. Tooling MUST support the <a href="#dialectSchemaId">OAS dialect schema id</a>, and MAY support additional values of `$schema`.

To allow use of a different default `$schema` value for all Schema Objects contained within an OAS document, a `jsonSchemaDialect` value may be set within the <a href="#oasObject">OpenAPI Object</a>. If this default is not set, then the OAS dialect schema id MUST be used for these Schema Objects. The value of `$schema` within a resource root Schema Object always overrides any default.

> For standalone JSON Schema documents that do not set `$schema`, or for Schema Objects in OpenAPI description documents that are _not_ [complete documents](#documentStructure), the dialect SHOULD be assumed to be the OAS dialect.

Reviewer 2: Just for my education: does this statement now define the desired implementation behavior in the OAS context?

And will this become a MUST in Moonwalk?

Author: These two lines aren't directly related, because I can't figure out what the old line meant.  The way we use git is so messed up that much of the past change history is obscured in a giant squashed mega-commit at the point when we released 3.1, so I can't figure out who wrote that text or why.  It was not part of the changes that I made for JSON Schema support in 3.1.

I have no idea why it would matter that you're coming _from_ some other JSON Schema document.

So it's possible that I'm changing something by removing that.  I'm just not sure what.  Maybe that you can't expect JSON Schema to know to use the OpenAPI dialect if the implementation doesn't know about OpenAPI?  Which is true, but the "JSON Schema rules" are that the behavior is implementation-defined, which means it can be anything and be valid according to those rules.  I'd be OK with leaving it in since AFAICT it's a no-op.  If you don't use `$schema` with regular JSON Schema then you potentially get random behavior anyway, so don't do that.

The SHOULD is just an interpretation of the `jsonSchemaDialect` guidance for what to do when `jsonSchemaDialect` is not set, but taking into account the fact that people write incomplete documents.  In this view, I'm treating a standalone JSON Schema document used by an OpenAPI Description as an incomplete OpenAPI Description document.

The SHOULD here has nothing to do with references _from_ JSON Schemas.
